src/main/java/org/jetbrains/bsp/bazel/server/bsp/resolvers/actiongraph/ActionGraphV2Parser.java

```diff
@@ -0,0 +1,109 @@
+package org.jetbrains.bsp.bazel.server.bsp.resolvers.actiongraph;
+
+import com.google.common.collect.Lists;
+import com.google.devtools.build.lib.analysis.AnalysisProtosV2;
+import java.util.*;
+import java.util.stream.Collectors;
+import org.jetbrains.bsp.bazel.commons.Uri;
+
+public class ActionGraphV2Parser implements ActionGraphParser {
+  private final AnalysisProtosV2.ActionGraphContainer actionGraph;
+  private final Map<Integer, AnalysisProtosV2.PathFragment> pathFragmentMap;
+
+  public ActionGraphV2Parser(AnalysisProtosV2.ActionGraphContainer actionGraph) {
+    this.actionGraph = actionGraph;
+    pathFragmentMap =
+        actionGraph.getPathFragmentsList().stream()
+            .collect(Collectors.toMap(AnalysisProtosV2.PathFragment::getId, fragment -> fragment));
+  }
+
+  @Override
+  public List<String> getInputsAsUri(String target, String execRoot) {
+    return getInputs(target, Lists.newArrayList(".jar", "js")).stream()
+        .map(exec_path -> Uri.fromExecPath(exec_path, execRoot).toString())
+        .collect(Collectors.toList());
+  }
+
+  private List<String> getInputs(String target, List<String> suffixes) {
+    return getActions(target).stream()
+        .flatMap(action -> action.getInputDepSetIdsList().stream())
+        .flatMap(
+            depset -> {
+              Queue<Integer> queue = new ArrayDeque<>();
+              queue.add(depset);
+              return expandDepsetToArtifacts(queue).stream();
+            })
+        .map(AnalysisProtosV2.Artifact::getPathFragmentId)
+        .map(pathFragmentId -> "exec-root://" + constructPath(pathFragmentId))
+        .filter(path -> suffixes.stream().anyMatch(path::endsWith))
+        .collect(Collectors.toCollection(TreeSet::new))
```

Comment:
I think you can use `disctinct()` here instead

Reply:
Will change in both parsers

Reply:
Well, I think that some logic should be extracted from both parsers to the separate class or to the interface